### ISSUE TITLE
Using `props` causes reactions and computeds to trigger in undesirable ways

### ISSUE THREAD

**Issue by iEchoic** [mobxjs/mobx-react#380]
Hello,

We've been using mobx reactions/computeds/etc. in what I thought was an idiomatic way, but I'm now finding that we're triggering a lot of unexpected reactions (and probably computed executions as well) when referencing `this.props`.  Here's the issue:

```
@observable someObservable = 'test';

// this autorun will change every time *any* props change, causing lots
// of excessive calls
autorun(() => {
  this.props.onChange(this.someObservable);
})

// this autorun will only change when `this.someObservable` or `this.props.onChange`
// changes
autorun(() => {
  const onChange = expr(() => this.props.onChange);
  onChange(this.someObservable);  
})
```

The only solution I see is to create a `computed` or use `expr` for every single prop usage.

I believe that this is behavior is by design in current mobx-react. If I'm understanding correctly, this is happening because `this.props` itself (as opposed to only the values on `props`) is observable, so every time props changes, it triggers reactions/computations. 

This behavior is unexpected to me. One of the reasons I love mobx is that if you write idiomatic mobx, everything magically computes/reacts a minimal number of times. However, the first example, this feels like an idiomatic way to use reactions and `prop`s to me, yet it behaves unexpectedly. Is there a better solution here, or some way we can make `props` with `autorun`/`computed`/etc. behave in a less-unexpected way?

**Comment by urugator** [comment]
It's impossible to make individual props observable, see #347
However I think the main problem here is that `onChange` handler is passed as anonymous inline function, therefore the "pure render" check always fails and props are always considered modified, which in turn triggers autoruns (to avoid staleness, when props are actually updated).

**Comment by iEchoic** [comment]
We never use inline anonymous functions for onChange handlers - it’s actually that other props are changing. I agree with @adamabeshouse in that thread that the ideal behavior would be that individual props are treated as observable (even if there’s some intermediate hack to make them appear to behave that way). Is the proposal at the end of that thread not possible/is “impossible” the right term here?

In #347 the example was somewhat non-idiomatic mobx, but the example above is, so I think this is a symptom that the observable behavior of props isn’t quite right. Is the issue purely an implementation problem, and we agree on the ideal behavior of individual props being observable? I can’t tell from that issue. I’d be interested in finding solutions to it if so.

**Comment by urugator** [comment]
> the ideal behavior of individual props being observable

Not entirely. Props are designed as immutable object in react, it doesn't make sense making immutable object observable, since only mutations can be observed.
You can't subscribe for an individual prop, because individual prop can't change. 
This is why you can't simply use `this.props.something` in render, but you have to use `this.props.observable.something`. The same limitation apply to any other reaction/autorun.
The thing is that if you use `this.props.something` the component is still updated, not because of MobX, but because component recieved new props. The same couldn't be said about custom `reaction/autorun` even if `this.props` was used inside. So `this.props` property was made observable to align the behavior of custom `autorun/reaction/computed` with the behavior of `render`. 
In other words, using `this.props` in `autorun` is a way to propagate component update caused by different means than MobX (an update caused by shallow prop change) to custom `autorun/reaction/computed` depending on such component update (depending on `this.props`).

**Comment by iEchoic** [comment]
Thanks for the explanation. 

> Props are designed as immutable object in react, it doesn't make sense making immutable object observable, since only mutations can be observed.

This doesn't prohibit us from having a facade `props` object with observable properties that are updated whenever the immutable `props` object changes, does it? 

> this.props property was made observable to align the behavior of custom autorun/reaction/computed with the behavior of render.

Note that if you align the behavior of `autorun`/`computed` with the behavior of `render`, then the user has to solve the problem mobx already solves for `render`, again: how to prevent extraneous renders when `props` change, except this time the user needs to solve it for `autorun`/`computed`, which are not beneficiaries of `PureComponent`/`shouldComponentUpdate`.

Current behavior aside, from a purely "how do we want this library to behave" perspective, do you think that:

```
autorun(() => {
  this.props.onChange(this.someObservable);
})
```

should result in `this.props.onChange` being called repeatedly with the same value even when neither `onChange` or `someObservable` changes? Or do we think that idiomatic mobx should look like this?:

```
autorun(() => {
  const onChange = expr(() => this.props.onChange);
  this.props.onChange(this.someObservable);
})
```

I doubt most mobx users (especially non-expert ones) would expect the first to behave the way it does, and it also seems like it would be a bad thing if idiomatic mobx required you to add `expr` or `computed` statements to every `prop` access like in the second example. 

The current behavior of `props` results in either a) confusing and inefficient behavior or b) lots of `expr`/`computed`s being littered around. This seems problematic to me. I'd be happy to help think of/contribute to solutions if it's agreed that this is currently a problem.

**Comment by urugator** [comment]
>  which are not beneficiaries of PureComponent/shouldComponentUpdate

But they are, the `autorun/reaction/computed` [is called only if props are shallowly modified](https://github.com/mobxjs/mobx-react/blob/master/src/observer.js#L142), exactly like `render`. 
Thats why I initially suggested that the extra calls are caused by inline function, which would always break the pure-render check.

> Or do we think that idiomatic mobx should look like this?:

The idiomatic way is not to rely on props in reactive functions, but on observables.
Again this is the very same as with `render` - even if you don't use some prop inside render and the prop is shallowly modified, the `render` still triggers.
From MobX perspective this can be considered inefficient behavior, that's why docs suggest to dereference values lately etc... the need for these suggestions also implies that the behavior can be *consfusing*...

**Comment by iEchoic** [comment]
> But they are, the autorun/reaction/computed is called only if props are shallowly modified, exactly like render.

Yes, but in vanilla react, `render` was the only thing that updated when props changed - so `shallowCompare` was effective at causing render only to render when something *used* by `render` actually changes. Now that (in mobx land) we have a bunch of things that also update based on props (`autorun`, `computed`, etc.) and are currently updating whenever *any* prop changes, we're now once again in a situation where we want to make those functions recompute only when something *used* changed.

> Thats why I initially suggested that the extra calls are caused by inline function, which would always break the pure-render check

Understood, but lots of things break the pure render check, like rendering `children`, so we can't point to props changing inefficiently as a cause.

> The idiomatic way is not to rely on props in reactive functions, but on observables.

This necessitates that simple onChange handlers - which are not things that are typically placed into stores and late-dereferenced, because they're component-level handlers and not data - are placed into stores or observables, which is unfortunate (imo). People aren't going to wrap every `onChange` handler in a store, so in practice this will mean that either people will have inefficient reactions or they'll discover this and add `expr`s all over.

Would just like to add this viewpoint/preference then, we can close this if it isn't something you're planning on changing.

**Comment by mweststrate** [comment]
@iEchoic I think your example can be solved more elegantly by using `reaction` to specify what you want to be reacting to specifically. The following wouldn't be reacting to new props

```javascript
reaction(() => this.someObservable, value => this.props.onChange(value))
```

We thought about updating individual fields instead of the entire props, however this deviates further from the react component model, where individual properties don't really change, rather you get a complete fresh set of properties. (an entire new `props` object). Of course we could perform diffing here and update the existing props based on the new props. The problem with that is that it doesn't really play nice with the lifecycle hooks of react components, that work with previous props, next props etc. 

Ideally, `.props` should not be observable *at all* (there are several open issues related to that). (But the problem then that is that @computed fields will keep using old properties)

**Comment by iEchoic** [comment]
@mweststrate thanks, that is definitely a better solution than `expr` for reactions.

FWIW, making props observable mostly solved the biggest mobx-react pain point for us, so I think that was a really great change.

To make a quick pitch for why I think individual field observability is the next big improvement over that: the only consistent pain point with mobx-react for us has been that components have to know too much about the observability of props being passed to them, making it difficult to create modular component interfaces while still encapsulating internal component logic. We want components to operate on *data*, not on *observability*. This was fixed with the observable props change, but now we can have inefficient computation based on our parent's behavior (unless we use `expr` very liberally, which is essentially re-creating individual field observability).

Pre-observable props:
```
@computed get checkedItems() {
  // this doesn't work when `items` changes if items is not an observable. We have to 
  // know if callers are using an observable to know if we can use this internal implementation, 
  // and potentially completely refactor this component based on that knowledge. This makes
  // `observable` a leaky abstraction

  return this.props.items.filter(items => item.isValid);
}
```

Current mobx-react:
```
@computed get checkedItems() {
  // Logically, this works now, and our component will behave correctly. We no longer have to
  // care about observability, only data. Great!
  // 
  // Except one parent is changing props a lot (perhaps changing children a lot)
  // so this is getting re-computed all the time and our perf mysteriously gets bad. 
  // Now we have to add `expr` for efficient computation, and once again we know and care  
  // too much about our parent behavior
  
  const items = expr(() => this.props.items);
  return items.filter(items => item.isValid);
}
```

If mobx-react had individually observable props:
```
@computed get checkedItems() {
  // mobx nirvana; we can treat `props` as data and everything works logically correctly and
  // efficiently. Our component can implement its internal logic however it wants, and 
  // we can create this component modularly, without knowledge about the behavior of its parent
  return this.props.items.filter(items => item.isValid);
}
```

Just want to add that perspective. Thanks!

**Comment by urugator** [comment]
>  only to render when something used by render actually changes. 

But thats not true, it doesn't matter if some prop is used or not, the render will re-run if ANY prop is shallowly modified.
The "re-runs only when used" behavior applies only to observables, but not to props itself.
The behavior of `render` and `autorun/computed/reaction/` is really the same. The only difference is that `render` depends on `this.props` implicitly, while `autorun/reaction/computed` depends on `this.props` explicitely (only when `this.props` is used), but both are updated under the same circumstances.

> if I'm making a checkbox that calls onChange, am I going to make a CheckboxStore?

If `onChange` represents state (changes over time), it must be stored as observable somewhere.
If you want to use MobX to update the component based on this state, you have to pass it as observable.
Otherwise the component is not updated via MobX and can't benefit from MobX tracking behavior.
You have to decide how you want to manage state. If you use a combination, the result will be efficient as the least efficient part (new props with shallow comparator in this case).
Some people suggest to have two kinds of components:
"dumb" - relies entirely on props, knows nothing about stores or observables, universally usable...
"smart" - connected to stores (observables), contains logic, domain specific...
(it seems unecessary to me)

> that is definitely a better solution than expr

Unlike the `expr` workaround, the proposed solution causes staleness when `onChange` is updated, but `this.someObservable` isn't. Functionally it's equivalent to:
```javascript
autorun(() => {
  const onChange = untracked(() => this.props.onChange);
  onChange(this.someObservable);
})
```
So it's not really a solution if you expect the same behavior.

**Comment by mweststrate** [comment]
>  the proposed solution causes staleness when onChange is updated

Good catch @urugator! Changes are indeed that calling the component with new props will also give it a new onChange handler (that really depends on how the component is called from it's parent).

In general I do not put reactions and such in components, but in stores which in general avoids these kind of issue at all. (I really try to keep my components dumb). If your component is simple, calling this.props.onChange directly from your event handlers rather then setting up a reaction for it is generally more easily to read. If it is complex, a view model is usually very powerful, and in that view model you can internally organize the reaction and just expose an api to attach / remove change handlers.

**Comment by iEchoic** [comment]
> "dumb" - relies entirely on props, knows nothing about stores or observables, universally usable...

I think this is the heart of the question. Should "dumb" components be able to utilize `@computed`/`autorun`? In the sense that those are very powerful constructs even for "dumb" components - and mobx is *way* more powerful as a whole if we answer "yes" - I feel like the answer should be "yes". If we answer "yes", then it seems clear that the current behavior of `props` in `computed`s and `autorun` on dumb components is not ideal (I can illustrate more with examples if this is unclear).

@mweststrate I think you're describing a viable way to navigate around these intricacies, but it seems to me that the fact that this way of thinking about component structure is so heavily dictated by the ramifications of how props update is an indicator that the current behavior could be improved. We originally used that pattern (lots of stores) and ended up running into limitations that caused us to move away from really granular component-level stores recently, and it's been a great move. I've been thinking about writing up our experiences/patterns we use with mobx, so I'll share a draft of that if I do incase you'd like to give thoughts. 

(p.s. we are still loving mobx, thanks for all of your work)

**Comment by urugator** [comment]
I think that the biggest optimization that MobX offers is the ability to update only a minimal amount of components (btw it wouldn't work well without shallow comparator). 
However if you decide to update your components by props, you will loose this ability, because prop updates must be propagated down through the tree by react (out of mobx control). Updating more components obviously leads to more reactions. 
By making props observable you would basically get "smarter" `shouldComponentUpdate` based on observables (more expensive than shallow comparator).
This would truly reduce the number of reactions and could stop the props propagation earlier (if the modified prop isn't used in render).
But the actual performance would highly depend on how much you rely on props, number of reactions and reaction's complexity.  
However it won't be better (or same) than directly relying on observables instead of props.

> not ideal

The interaction between observables and props isn't ideal, there is no denial. 
But I am (speaking for myself) not convinced that making them observable is a way to go for the reason stated above:
- how does it relate to react (in theory/practice)?
- how does it actually perform? 
- how to implement it? (I believe the Proxy support is required)
- does the complexity worth it, if problem can be (easily?) avoided while providing better performance at the same time?

**Comment by iEchoic** [comment]
@urugator I agree and that all makes sense, thanks for the explanation. I think that observable props are more correct semantically, but I don't have answers to bullets #2 and #3 and am not intimately familiar with mobx' internals, so admittedly I don't know if this is the right way to go either in practice. If you ever end up testing this, I'd love to try out an early version and help benchmark.

**Comment by domehead100** [comment]
Using mobx in the ways most commonly demonstrated involves creating some store object with an observable property or properties and then passing that object to observer components so that they can react to changes in the observable properties of the object (or possibly mutate the properties so that other reactions occur).

This can create a bit of a challenge for building reusable observer components because the component needs to know the shape (property name(s)) of the store object to use its observable properties.  This creates a very explicit dependency on that particular store object (or one shaped exactly like it, with the same property names).  This dependency can be quite extreme if the object passed to the observer components is the root store itself and the property of interest is several "dots" in from the root (a newbie mistake, which I myself made when starting out with mobx).  Of course, ```@inject``` can help to limit the depth of the "dots" by passing in specific store objects instead of the entire store.

As @urugator  says in #404:
> In other words you should always use this.props.somethingObservable.something instead of just this.props.something.

In that example, a component when passed ```somethingObservable``` as a prop would need to know that the name of the property to reference is ```something```.  Let's say that later you want to reuse the same component against a property ```someProperty``` on some other store object (let's say both "something" and "someProperty" are booleans).   Now your component won't work out of the box because it was hard-coded to look for a property called "something" rather than "someProperty."

Something (pun) I've been experimenting with at work is to formalize the interface for observable props in reusable observer components.  We use typescript, so the reusable component declares the property as type IRefProp<some type> and then in JSX it becomes very clear via intellisense when passing props to a component that an IRefProp<some type> is needed, and so we use either the ref() or rref() function shown in the code excerpt at the bottom of this post when passing the prop.  All IRefProps have a single property named "value", so the component does not need to know the shape of the observable in the actual store or model.  ref() and rref() are factory functions that return a RefProp or RefPropReadOnly.

Here's what a reusable observer component might look like:

```javascript
@observer2
class ChildComponent extends React.Component<{
someProp: IRefProp<boolean>;
anotherProp: string;
}> {
render() {
  const { anotherProp, someProp } = this.props
  <span>{anotherProp}</span>
  {someProp.value
    ? <span>Something rendered if someProp is true</span>
    : <span>Something rendered if someProp is false</span>
  }
}
```
This should re-render if anotherProp changes or if someProp.value changes.  I'm not showing it here, but there could be computeds or other reactions in the component based only on someProp.value, and if someProp.value is not used in the render, then no re-render would occur but the computeds should recalculate appropriately (I think).

And here's how we would pass the ref prop to the component:

```javascript
// store.checks is an object that contains an observable property someProp that is boolean
<ParentComponent>
  <ChildComponent someProp={ref(store.checks, "someProp")} />
</ParentComponent>
```
This is almost great, except for two things:  1) it creates a new RefProp in every render of the parent because the ref() function returns new RefProp, and thus 2) it suppose it causes unnecessary renders in the child because of the shallow compare of all props.  I've mitigated the latter by tweaking the observer code (calling it observer2 at this point) to check for instanceof RefProp and then for the RefProps do the shallow compare on the obj property instead of the RefProp itself. 

I'm also looking into making the ref function a static factory method on RefProp and then memoize its result using a WeakMap to cache instances so that a new RefProp is not created if one already exists for the same obj+propName.  This might also work nicely if the same store object+propName is passed to multiple components so that they share the same RefProp as well as the same store object.

Basically though, I think this is a fairly clean method of separating out the "observable" props from the rest of the props while making it slightly more straightforward to create reusable observer components.  If done properly, we should get the desired reactions to the observable props without extra renders (I think) and without losing renders due to non-observable props changing.

Wondering what others think about all of this.

Here's my current implementation of RefProp:

```javascript
import { observable } from "mobx";

export interface IRefProp<T> {
    value: T;
}

export interface IRefPropReadOnly<T> {
    readonly value: T;
}

export class RefProp<T> implements IRefProp<T> {
    @observable public obj: object;
    public propName: string | number;

    constructor(obj: object, propName: string | number) {
        this.obj = obj;
        this.propName = propName;
    }

    get value() {
        return this.obj[this.propName] as T;
    }

    set value (newValue: T) {
        this.obj[this.propName] = newValue;
    }
}

// tslint:disable-next-line:max-classes-per-file
export class RefPropReadOnly<T> implements IRefPropReadOnly<T> {
    @observable public obj: object;
    public propName: string | number;

    constructor(obj: any, propName: string | number) {
        this.obj = obj;
        this.propName = propName;
    }

    get value() {
        return this.obj[this.propName] as T;
    }

    // tslint:disable-next-line:no-unused
    set value(newValue: T) {
        throw new Error(`Error: tried to mutate the read-only read prop ${this.propName}`);
    }
}

/**
 * Reference prop - takes an object and a property name and creates a wrapper that
 * allows the property to be observable.  In this version, as opposed to rref, the 
 * property can be mutated by the receiver of the prop.
 * @param obj - object that has the property that we want to observe
 * @param propName - the string or number key of the property we want to observe
 */
export function ref<T>(obj: object, propName: string | number): IRefProp<T> {
    return new RefProp<T>(obj, propName);
}

/**
 * Read-only reference prop - takes an object and a property name and creates a wrapper that
 * allows the property to be observable.  This is the read-only version that forbids the receiver
 * of the ref prop from mutating it; though it can still observe it.
 * @param obj - object that has the property that we want to observe
 * @param propName - the string or number key of the property we want to observe
 */
export function rref<T>(obj: object, propName: string | number): IRefProp<T> {
    return new RefProp<T>(obj, propName);
}
```

**Comment by urugator** [comment]
@domehead100 

You may also consider creating references in stores: `@observable prop = observable.box(5)` (would be smart to use some store factory to enforce this as a rule and save some typing).
It all comes down to the fact that values can't be observed. Therefore all `observers/reactions/computed/autorun` must depend on some abstraction instead of values directly. 
The smallest abstraction possible is some "value reference" (a box in Mobx).
So if you wrap every value in this box, you will get a lot more flexibility when designing `observers` - their dependencies can stay focused on things they really need.
Btw it's also useful in stores/models, you don't need to pass the whole store as depedency, only the required parts.  
Obviously there is a cost - increased complexity, verbosity, `a.get().b.get().c.get()` and redundancy (more "observable boxes" than actually required). Also default conversion to observables won't work. 

> it creates a new RefProp

I wouldn't bother, any caching could end up being more expensive than a creation of a few additional objects. 

Btw it's possible to keep "the dereferencing part" inside the parent, while subscribing only the child:  
```javascript
<ParentComponent>
  <Observer>
    {_ => {
      <ChildComponent checked={store.checks.someProp}/>
    }}
  </Observer>
<ParentComponent>
```

**Comment by zocky** [comment]
I see that this issue is getting some real progress now, but let me chime in with the pattern that I know use to achieve the desired behavior:

````js
class TestValue extends React.Component {
  @observable showValue = false;
  @observable value = 10;

  constructor(props, context) {
    super(props, context);
    this.assignObservableProps(props);
  }
  componentWillReceiveProps(props, context) {
    this.assignObservableProps(props);
  }

  @action
  assignObservableProps(props) {
    this.value = props.value;
    this.showValue = props.showValue;
  }

  shouldComponentUpdate(props, context) {
    return false; // to prevent updates from changes to props
  }

  render() {
    console.log("render"); // will not happen if value changes while showValue is false
    if (this.showValue) return <div>{this.value}</div>;
    else return "Nothing shown";
  }
};
````

Now if only I knew how to make a decorator that would automate this for me ...

**Comment by zocky** [comment]
so I tried to write the decorator for the above, and here goes:

```jsx
function observerWithProps(oProps) {
  return function(Class) {
    return (props, context) => {
      const instance = observer(new Class(props, context));
      const keys = Object.keys(oProps);
      keys.forEach(key => {
        extendObservable(instance, {
          [key]: observable
        });
        instance[key] = key in props ? props[key] : oProps[key];
      });

      const componentWillReceiveProps = instance.componentWillReceiveProps;
      instance.componentWillReceiveProps = action(function(props, context) {
        componentWillReceiveProps &&
          componentWillReceiveProps.call(instance, props, context);
        keys.forEach(oProp => {
          instance[oProp] = props[oProp];
        });
      });

      instance.shouldComponentUpdate = function(props, context) {
        return false;
      };
      return instance;
    };
  };
}
```
So the above example turns into:

``` jsx
@observerWithProps({
  showValue: false,
  value: 10
})
class TestValue extends React.Component {
  render() {
    console.log("render"); // will not happen if value changes while showValue is false
    if (this.showValue) return <div>{this.value}</div>;
    else return "Nothing shown";
  }
};
```

**Comment by FredyC** [comment]
Haven't read all comments, but given the age of this issue, I will close it. If there is still something unresolved, let me know.
